pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -454,11 +500,82 @@ func (lb *ListenerBuilder) getListeners() []*listener.Listener {
 	return lb.gatewayListeners
 }
 
-// Create pass through filter chains matching ipv4 address and ipv6 address independently.
-// This function also returns a boolean indicating whether or not the TLS inspector is needed
-// for the filter chain.
-func buildInboundCatchAllNetworkFilterChains(configgen *ConfigGeneratorImpl,
-	node *model.Proxy, push *model.PushContext) ([]*listener.FilterChain, bool) {
+// getFilterChainMatchOptions returns the FilterChainMatchOptions that should be used based on mTLS mode and protocol
+func getFilterChainMatchOptions(settings plugin.MTLSSettings, protocol istionetworking.ListenerProtocol) []FilterChainMatchOptions {
+	switch protocol {
+	case istionetworking.ListenerProtocolHTTP:
+		switch settings.Mode {
+		case model.MTLSStrict:
+			return inboundStrictHTTPFilterChainMatchOptions
+		case model.MTLSPermissive:
+			return inboundPermissiveHTTPFilterChainMatchWithMxcOptions
+		default:
+			return inboundPlainTextHTTPFilterChainMatchOptions
+		}
+	case istionetworking.ListenerProtocolAuto:
+		switch settings.Mode {
+		case model.MTLSStrict:
+			return inboundStrictFilterChainMatchOptions
+		case model.MTLSPermissive:
+			return inboundPermissiveFilterChainMatchWithMxcOptions
+		default:
+			return inboundPlainTextFilterChainMatchOptions
+		}
+	default:
+		switch settings.Mode {
+		case model.MTLSStrict:
+			return inboundStrictTCPFilterChainMatchOptions
+		case model.MTLSPermissive:
+			return inboundPermissiveTCPFilterChainMatchWithMxcOptions
+		default:
+			return inboundPlainTextTCPFilterChainMatchOptions
+		}
+	}
+}
+
+type fcOpts struct {
+	matchOpts FilterChainMatchOptions
+	fc        istionetworking.FilterChain
+}
+
+func (opt fcOpts) populateFilterChain(mtls plugin.MTLSSettings, port uint32, matchingIP string) fcOpts {
+	opt.fc.FilterChainMatch = &listener.FilterChainMatch{}
+	opt.fc.FilterChainMatch.ApplicationProtocols = opt.matchOpts.ApplicationProtocols
+	opt.fc.FilterChainMatch.TransportProtocol = opt.matchOpts.TransportProtocol
+	if len(matchingIP) > 0 {
+		opt.fc.FilterChainMatch.PrefixRanges = []*core.CidrRange{util.ConvertAddressToCidr(matchingIP)}
+	}
+	if port > 0 {
+		opt.fc.FilterChainMatch.DestinationPort = &wrappers.UInt32Value{Value: port}
+	}
+	opt.fc.ListenerProtocol = opt.matchOpts.Protocol
+	if opt.fc.ListenerProtocol == istionetworking.ListenerProtocolHTTP {
+		opt.fc.TLSContext = mtls.HTTPTLSContext
+	} else {
+		opt.fc.TLSContext = mtls.TCPTLSContext
+	}
+	return opt
+}
+
+func getMtlsSettings(configgen *ConfigGeneratorImpl, in *plugin.InputParams, passthrough bool) []plugin.MTLSSettings {
+	for _, p := range configgen.Plugins {
+		cfg := p.InboundMTLSConfiguration(in, passthrough)
+		if cfg != nil {
```

Comment:
this means we will always use the result from the first plugin, currently only the authn plugin use this and after this I feel it's even less useful to leave this in the plugin interface? Maybe this interface should be removed from the plugin completely later.

Reply:
I agree, we discussed in networking WG removing the plugin interface entirely. But this PR was already pretty big so I would prefer to tackle it later?

Reply:
yes, it should be done in a separate PR.